File: 2-add-two-numbers/2-add-two-numbers.scala
object Solution {
    def addTwoNumbers(l1: ListNode, l2: ListNode): ListNode = {
        var i = l1
        var j = l2
        var c = 0

        val res = ListNode()
        var head = res

        while(i != null || j != null) {
            val vi = if (i != null) {
                val v = i.x
                i = i.next
                v
            } else 0

            val vj = if (j != null) {
                val v = j.x
                j = j.next
                v
            } else 0

            head.next = ListNode((vi + vj + c) % 10)
            c = (vi + vj + c) / 10

            head = head.next
        }

        if (c != 0) head.next = ListNode(c)

        res.next
    }
}

File: 2-add-two-numbers/2-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode res = new ListNode();
        ListNode current = res;
        
        int c = 0;
        while(l1 != null || l2 != null) {
            int v1 = (l1 == null)? 0 : l1.val;
            int v2 = (l2 == null)? 0 : l2.val;
            
            int sum = v1 + v2 + c;
            c = sum / 10;
            
            current.next = new ListNode(sum % 10);
            current = current.next;
            
            if (l1 != null) l1 = l1.next;
            if (l2 != null) l2 = l2.next;
        }
        
        if (c != 0) current.next = new ListNode(c);
        
        return res.next;
    }
}